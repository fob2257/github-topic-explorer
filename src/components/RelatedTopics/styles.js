import styled from 'styled-components';

export const RelatedTopicsDiv = styled.div`
  margin-top: 32px;
`;

export const Heading = styled.h2`
  margin: 0;
  font-weight: 400;
  color: #263238;
  font-size: 24px;
`;
