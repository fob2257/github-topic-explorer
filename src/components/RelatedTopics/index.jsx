import React, { useState } from 'react';

import { RelatedTopicsDiv, Heading } from './styles';
import RelatedTopic from '../RelatedTopic';

const RelatedTopics = ({ topics }) => {
  const [relatedTopicId, setRelatedTopicId] = useState(null);

  const toggleRelatedTopic = (topic) => {
    setRelatedTopicId(topic.id === relatedTopicId ? null : topic.id);
  };

  return (
    <RelatedTopicsDiv>
      <Heading>Related topics: {topics.length}</Heading>
      {topics.map((topic) => (
        <RelatedTopic
          key={topic.id}
          topic={topic}
          expanded={topic.id === relatedTopicId}
          toggleRelatedTopic={toggleRelatedTopic}
        />
      ))}
    </RelatedTopicsDiv>
  );
};

export default RelatedTopics;
