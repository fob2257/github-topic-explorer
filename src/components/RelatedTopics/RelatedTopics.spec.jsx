import * as React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { create } from 'react-test-renderer';

import RelatedTopics from './RelatedTopics';

const topics = [
  {
    __typename: 'Topic',
    id: 'MDU6VG9waWNyZWFjdC1uYXRpdmU=',
    name: 'react-native',
    stargazerCount: 22240,
    relatedTopics: [
      {
        __typename: 'Topic',
        id: 'MDU6VG9waWNyZWFjdGpz',
        name: 'reactjs',
        stargazerCount: 969,
      },
    ],
  },
  {
    __typename: 'Topic',
    id: 'MDU6VG9waWNhbmd1bGFy',
    name: 'angular',
    stargazerCount: 39188,
    relatedTopics: [
      {
        __typename: 'Topic',
        id: 'MDU6VG9waWN2dWU=',
        name: 'vue',
        stargazerCount: 45009,
      },
    ],
  },
];

describe('<RelatedTopics />', () => {
  const setup = () => <RelatedTopics topics={topics} />;

  it('should render correctly', () => {
    render(setup());

    const element = screen.getByRole('heading', { name: /related topics/i });

    expect(element).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const component = create(setup());
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should dispatch setRelatedTopicId if topic is clicked', async () => {
    const dispatch = jest.fn();
    const useStateSpy = jest.spyOn(React, 'useState');
    useStateSpy.mockReturnValue([false, dispatch]);

    render(setup());

    const element = screen.getByText(new RegExp(topics[0].name, 'i'));

    expect(element).toBeInTheDocument();

    fireEvent.click(element);

    const calls = dispatch.mock.calls.flat();

    expect(calls.length).toBeGreaterThan(0);

    useStateSpy.mockRestore();
  });
});
