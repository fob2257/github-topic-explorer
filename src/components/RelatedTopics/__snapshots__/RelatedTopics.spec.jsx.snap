// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RelatedTopics /> should match snapshot 1`] = `
<div
  className="sc-bdvvtL bIXkDV"
>
  <h2
    className="sc-gsDKAQ jJCLbz"
  >
    Related topics: 
    2
  </h2>
  <div
    className="sc-dkPtRN caPTMa"
    onClick={[Function]}
  >
    <h3
      className="sc-hKwDye hoxHUg"
    >
      react-native
    </h3>
    <div
      className="sc-gKclnd fVLuFs"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
      <span
        className="sc-iCfMLu llrDsR"
      >
        22240
      </span>
    </div>
  </div>
  <div
    className="sc-dkPtRN caPTMa"
    onClick={[Function]}
  >
    <h3
      className="sc-hKwDye hoxHUg"
    >
      angular
    </h3>
    <div
      className="sc-gKclnd fVLuFs"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="StarIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
        />
      </svg>
      <span
        className="sc-iCfMLu llrDsR"
      >
        39188
      </span>
    </div>
  </div>
</div>
`;
